[BREAKINGCHANGE] Define Queries in PanelSpec and Implement DataQueriesProvider

eunicorn opened the pull request:
There are a few missing features that we would like support in Perses Panel:

- Ability to have standard loading and error states across all panels
- Ability to display query details in Panel Header 
  - show template variables used in queries
  - show if queries are truncated 

For example:
![image](https://user-images.githubusercontent.com/9817826/228094675-25d7eeef-a75a-4894-8d16-d7c0a3acc21d.png)

In order to support these missing features, we need to...

1. Move query definitions from panel plugin up a level to where the panel spec is defined
2. Implement `DataQueriesProvider` that is responsible for running queries so that panel plugins are no longer responsible for data querying.
3. Implement `useDataQueries` hook that panel plugins can call and retrieve the query results.

See [discussion here](https://github.com/perses/perses/discussions/956#discussioncomment-4988816). 

FUTURE WORK:

- update migration script
- decouple query editor from panel plugin spec

eunicorn commented on ui/panels-plugin/src/model/index.ts:
```diff
@@ -0,0 +1,14 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export * from './time';
```
exporting `useSuggestedStepMs` here

eunicorn commented on ui/core/src/model/query.ts:
```diff
@@ -0,0 +1,25 @@
+  plugin: Definition<PluginSpec>;
+}
+
+export interface QueryDefinition<Kind = QueryType, PluginSpec = UnknownSpec> {
```
A generic query definition interface that can extend to support more than just `TimeSeriesQuery`

eunicorn commented on ui/dashboards/src/components/PanelDrawer/PanelEditorForm.tsx:
```diff
@@ -65,10 +67,17 @@ export function PanelEditorForm(props: PanelEditorFormProps) {
   };
 
+  // TO DO: once we remove query editor from panel plugin, we can remove this method
+  const handlePluginSpecChange: OptionsEditorProps<UnknownSpec>['onChange'] = (spec) => {
```
Since the query editor is part of the panel plugin, the queries are populated in plugin spec every time you update it the query editor. Thus we need to keep track of the state of the `queries`

eunicorn commented on ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartQueryEditor.tsx:
```diff
@@ -43,7 +42,11 @@ export function TimeSeriesChartQueryEditor(props: TimeSeriesChartOptionsEditorPr
     onChange(
       produce(value, (draft: TimeSeriesChartOptions) => {
-        draft.queries[index] = queryDef;
+        if (draft.queries) {
```
need to check if `draft.queries` exists because we don't define the `queries` in the plugin spec anymore

eunicorn commented on ui/plugin-system/src/runtime/DataQueriesProvider/DataQueriesProvider.tsx:
```diff
@@ -0,0 +1,76 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { createContext, useCallback, useContext, useMemo } from 'react';
```
Decided to not use `use-context-selector` here because all context values are updated every time the query changes so even if we use selector, all components will be triggered to update anyways

juliepagano replied:
I think we may benefit from some perf tuning around context, but that'll be better to do as a follow-up in the future. Agreed with not worrying about it in this particular PR.

eunicorn replied:
Agreed, I was talking to steven and mentioned that TimeRangeProvider might be a good candidate for `use-context-selector` because not all values are updated at the same time. For instance, absolute time range doesn't always change even when you refresh the time range

eunicorn commented on ui/dashboards/src/utils/time.ts:
moved `useSuggestedStepMs` from panels-plugin to dashboards because we now call it in `GridItemContent`

eunicorn commented on ui/dashboards/src/utils/time.ts:
moved `useSuggestedStepMs` from panels-plugin to dashboards because we now call it in GridItemContent

eunicorn commented on ui/dashboards/src/utils/time.ts:
moved `useSuggestedStepMs` from panels-plugin to dashboards because we now call it in GridItemContent

sjcobb replied:
I’m still making my way through this branch, but calling it directly in GridItemContent seems a little counterintuitive. Would some new wrapper component make sense? Still thinking of what to call it though 🤔

eunicorn replied:
Hmm i thought about making a wrapper component like `DataQueryingPanel` but is it confusing to have both `Panel` and `DataQueryingPanel`? Also if we decide to have `DataQueryingPanel`, `useSuggestedStepMs` will still live in the `dashboard` package so I don't really see much benefit adding a wrapper component here

eunicorn commented on pkg/model/api/v1/dashboard.go:
@Nexucis FYI updated `PanelSpec` to include `Queries`

eunicorn replied:
tagging @AntoineThebaud as well since this will break the migration script

AntoineThebaud replied:
regarding the dashboard migration, this basically implies to remove the existing `query` (or `queries` for timeseries) mapping logic from the individual `mig.cuepart` file of each plugin, and add a new logic to the main/central `mapping.cuepart` ([within this `spec`](https://github.com/perses/perses/blob/main/internal/api/shared/migrate/mapping.cuepart#L86)). The code should look very much like what we have currently in the [timeseries plugin](https://github.com/perses/perses/blob/main/schemas/panels/time-series/mig.cuepart#L4) since we already had a queries array for this one.

sjcobb commented on pkg/model/api/v1/dashboard.go:
```diff
@@ -27,13 +27,23 @@ import (
 	Spec PanelSpec `json:"spec" yaml:"spec"`
 }
 
+type Query struct {
```
Did a quick test by adding a new dashboard, adding a TimeSeriesChart panel with an `up` query, and onSave it resulted in: 
<img width="843" alt="image" src="https://user-images.githubusercontent.com/9369625/228343508-c8e99f1c-0fd6-4cc0-893c-317155a7f705.png">

Do you think there will be some additional work on the BE side besides what you mentioned around migration?

eunicorn replied:
I am able to save a new dashboard with time series panel with no issue. Are you doing something different than what I did in the video?

https://user-images.githubusercontent.com/9817826/228346184-f7bbb105-f140-4aa4-9884-00b0423eda98.mov

sjcobb replied:
Oh wait, duh I need to rebuild the BE again, I’ll try it again in a little bit but I forgot to do that 🙃

sjcobb replied:
Okay testing this locally again and yeah it works! Notice a few small things like seeing this warning in the console:

![image](https://user-images.githubusercontent.com/9369625/228850583-37c639fd-1caa-4aa6-8f44-cc6105ff0a5e.png)

sjcobb replied:
Having trouble reproducing the can't add a threshold issue now 🤔, the JSON tab not being accurate to where queries moved in panel spec is throwing me off, now I added a few panels, renamed one, and went to save and got:
<img width="663" alt="image" src="https://user-images.githubusercontent.com/9369625/228855825-010fe853-de04-4c12-871e-3986a9c9552f.png">

Here is the JSON I ended up with that wouldn't save:
```json
{ "kind": "Dashboard", "metadata": { "name": "QueriesSpecRefactor", "created_at": "2023-03-28T19:11:01.319754Z", "updated_at": "2023-03-30T13:21:00.259488Z", "version": 6, "project": "perses" }, "spec": { "display": { "name": "QueriesSpecRefactor" }, "panels": { "Test1": { "kind": "Panel", "spec": { "display": { "name": "Test 1" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_duration_seconds" } } } } ], "legend": { "position": "Right" }, "thresholds": { "steps": [ { "value": 0.01 }, { "color": "#FFB249", "value": 0.05 } ] }, "visual": { "line_width": 3, "area_opacity": 0, "point_radius": 2.5, "connect_nulls": false } } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_duration_seconds" } } } } ] } }, "Test1-1": { "kind": "Panel", "spec": { "display": { "name": "Test 1" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "Test1-2": { "kind": "Panel", "spec": { "display": { "name": "Test 1" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "Test1-3": { "kind": "Panel", "spec": { "display": { "name": "Test 1" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "up" } } } } ] } }, "Test1-4": { "kind": "Panel", "spec": { "display": { "name": "Test 1" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_duration_seconds" } } } } ] } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_duration_seconds" } } } } ] } }, "": { "kind": "Panel", "spec": { "display": { "name": "Test" }, "plugin": { "kind": "TimeSeriesChart", "spec": { "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_http_content_length", "series_name_format": "{{ instance }}" } } } } ], "legend": { "position": "Bottom" }, "visual": { "line_width": 1.5, "area_opacity": 0.6, "point_radius": 4, "connect_nulls": true }, "y_axis": { "show": true, "label": "Test", "unit": { "kind": "Decimal", "decimal_places": 2, "abbreviate": false } }, "thresholds": { "steps": [] } } }, "queries": [ { "kind": "TimeSeriesQuery", "spec": { "plugin": { "kind": "PrometheusTimeSeriesQuery", "spec": { "query": "probe_http_content_length", "series_name_format": "{{ instance }}" } } } } ] } } }, "layouts": [ { "kind": "Grid", "spec": { "display": { "title": "Panel Group", "collapse": { "open": true } }, "items": [ { "x": 12, "y": 6, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/Test1" } }, { "x": 12, "y": 0, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/Test1-1" } }, { "x": 0, "y": 0, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/Test1-4" } }, { "x": 0, "y": 6, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/Test1-2" } }, { "x": 0, "y": 12, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/" } } ] } }, { "kind": "Grid", "spec": { "display": { "title": "Second Group", "collapse": { "open": true } }, "items": [ { "x": 0, "y": 0, "width": 12, "height": 6, "content": { "$ref": "#/spec/panels/Test1-3" } } ] } } ], "variables": [], "duration": "5m" } }
```

sjcobb replied:
But again, when I start fresh with a brand new dashboard, adding new panels and saving works 😕 😕 

I guess I somehow got one panel into a bad state before and then that happened to be the one I duplicated a bunch before trying to save

eunicorn replied:
I'm working on moving the whole panel definition to the JSON tab so ideally that should fix this issue!

eunicorn replied:
WIP but most of the work is done on this branch [https://github.com/perses/perses/tree/eunicorn/decouple-query-editor](https://github.com/perses/perses/tree/eunicorn/decouple-query-editor) if you want a sneak peak 👀

eunicorn replied:
I cannot reproduce the threshold issue in this branch or [eunicorn/decouple-query-editor](eunicorn/decouple-query-editor) so marking this conversation as resolved! feel free to ping me or comment again if you see another issue :)

sjcobb commented on ui/dashboards/src/components/Panel/Panel.tsx:
```diff
@@ -28,7 +28,7 @@ export interface PanelProps extends CardProps<'section'> {
  * Renders a PanelDefinition's content inside of a Card.
  */
-export function Panel(props: PanelProps) {
+export const Panel = memo(function Panel(props: PanelProps) {
```
🎉 good find!

juliepagano commented:
> * update migration script
> * decouple query editor from panel plugin spec **(or put the full JSON into the JSON tab like we discussed earlier today)**

Is it safe to do these follow-ups separately or are we potentially going to leave things in a broken state if someone does a release before those two get fixed? If the latter, it might be better to work this out in a feature branch, so you can balance having several PRs without shipping perses in a potential state that breaks something.

juliepagano commented:
I _think_ those e2e tests might be failing because the edit panel is still open, so the page finds two "simple stat" panels. If you wait for the side panel to finish closing in the tests when you're done editing, I _think_ it'll fix the issue. I think the "panel editor" e2e helper has an "isClosed" method you can use for that.

juliepagano commented on ui/dashboards/src/components/Panel/Panel.stories.tsx:
```diff
@@ -0,0 +1,111 @@
+export const EditMode: StoryFn<typeof Panel> = (args) => {
+  const editHandlers = {
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    onEditPanelClick: () => {},
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    onDuplicatePanelClick: () => {},
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    onDeletePanelClick: () => {},
```
Could you make these `actions`, so that they show up in the little actions pane in storybook? You can find something similar in the stories for `EChart`. You shouldn't need the eslint-disables after doing that.

https://github.com/perses/perses/blob/main/ui/components/src/EChart/EChart.stories.tsx#L63-L69

juliepagano commented on ui/dashboards/src/components/Panel/Panel.stories.tsx:
```diff
@@ -0,0 +1,111 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Thank you for adding stories for this! It's cool to see how a panel works by itself like this. 😄

eunicorn replied:
It was cool and easy to see what providers a panel needs! The only issue I noticed is when you click "Show code" on the docs page, the sample code is not very helpful.

juliepagano commented on ui/dashboards/src/components/Panel/Panel.stories.tsx:
```diff
@@ -0,0 +1,111 @@
+    WithDataQueries,
+    WithTemplateVariables,
+    WithTimeRange,
+    WithDatasourceStore,
```
Note that this is currently set up to query the prometheus demo site, which will not be reliably consistent for visual testing. You can either: disable perses for stories that use that OR configure the store to use mock data like we do in some of our e2e tests. Considering this PR is already pretty big, I'd recommend the first one, and we can circle back on the mock data later. 

You can find an example of how to disable happo in https://github.com/perses/perses/blob/main/ui/dashboards/src/components/Dashboard/Dashboard.stories.tsx#L177

juliepagano commented on ui/dashboards/src/components/PanelDrawer/PanelEditorForm.tsx:
```diff
@@ -65,10 +67,19 @@ export function PanelEditorForm(props: PanelEditorFormProps) {
+  // TO DO: once we remove query editor from panel plugin, we can remove this method
+  const handlePluginSpecChange: OptionsEditorProps<UnknownSpec>['onChange'] = (spec) => {
+    // time series chart query editor returns spec.queries while gauge and stat charts return spec.query
+    if (spec?.queries || spec?.query) {
+      setQueries((spec?.queries ?? [spec?.query]) as QueryDefinition[]);
+    }
```
Are there any edge cases you're missing where someone intentionally sets the queries to something falsy? (Not a leading question -- genuinely curious because I can't remember all the logistics of this code).

juliepagano commented on ui/dashboards/src/components/PanelDrawer/PanelPreview.tsx:
```diff
@@ -34,9 +37,30 @@ export function PanelPreview({ name, description, kind, spec }: PanelEditorValue
+  // ex: markdown panel does not have a query
+  if (!queries) {
+    return (
+      <Box height={300}>
```
Can this height value get pulled out into a constant now that you're using it in multiple places?

juliepagano commented on ui/dashboards/src/context/DashboardProvider/DashboardProvider.tsx:
```diff
@@ -80,8 +80,26 @@ export function DashboardProvider(props: DashboardProviderProps) {
+      initialValues: {
+        kind: defaultPanelKind,
+        spec,
+        queries: [
+          {
+            kind: 'TimeSeriesQuery',
+            spec: {
+              plugin: {
+                kind: 'PrometheusTimeSeriesQuery',
+                spec: {
+                  query: '',
+                },
+              },
+            },
+          },
+        ],
```
Should this default be configurable like `defaultPanelKind` or an empty array? Will all dashboard users always have the `PrometheusTimeSeriesQuery` kind with how plug-and-play this is with plugins?

eunicorn replied:
updated this to check `defaultPluginKinds`, similar to how it currently works with `defaultPanelKind`

juliepagano commented on ui/dashboards/src/stories/decorators/WithDataQueriesProvider.tsx:
```diff
@@ -0,0 +1,41 @@
+): parameter is WithDataQueriesProvideParameter {
+  return !!parameter && typeof parameter === 'object' && 'props' in parameter;
+}
+
```
Thank you for adding another provider helper! 🎉

juliepagano commented on ui/e2e/src/tests/statChartPanel.spec.ts:
```diff
@@ -61,7 +61,7 @@ test.describe('Dashboard: Stat Chart Panel', () => {
         await page.keyboard.press('Escape');
       });
-      const panel = dashboardPage.getPanelByName('Simple Stat');
+      const panel = dashboardPage.getPanelByName('Simple Stat', { nth: 0 });
```
Left you a comment at the bottom of the PR about what I _think_ might be happening here. My guess is this was a preexisting potentially flaky issue and something about your change increased the likelihood of it happening -- I can't reproduce locally, but I could guess at the "why" when looking at the tests.

eunicorn replied:
Actually it was because i mislabeled something in StatChartPanel.tsx so it was actually rendering two StatCharts. The test is now passing again :)

juliepagano reviewed (commented):
First skim on the code. I'll take another pass later today or early tomorrow.

saminzadeh commented on ui/plugin-system/src/runtime/DataQueriesProvider/DataQueriesProvider.tsx:
```diff
@@ -0,0 +1,76 @@
+import { useTimeSeriesQueries } from '../time-series-queries';
+import { DataQueriesProviderProps, QueryData, UseDataQueriesResult } from './model';
+
+export function useDataQueries(): UseDataQueriesResult {
```
Nit: I wonder if we should rename this to be more clear that it returns the data?
```
useQueriesData()
useQueryDataResults()
```

saminzadeh commented on ui/plugin-system/src/runtime/DataQueriesProvider/DataQueriesProvider.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  const ctx = useMemo(() => {
+    return {
+      queryResults: data,
```
Would a top level `errors` object be helpful? I suppose we can always add it in.

saminzadeh commented on ui/dashboards/src/components/GridLayout/GridItemContent.tsx:
```diff
@@ -37,5 +45,26 @@ export function GridItemContent(props: GridItemContentProps) {
+      };
+    });
+
+    return (
```
I think this code is fine. It's more about how we setup the plugins and what they export. Ideally, we export the context aware component in addition to the [panel plugin](https://github.com/perses/perses/blob/433cc551b2be7cad1a5757c0240d7149a7d1e933/ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChart.ts#L24). That means, there really should be two exports:

```tsx
import { TimeSeriesChartPanel, TimeSeriesChart } from "@perses-dev/panel-plugins"
```

`TimeSeriesChartPanel` speaks  JSON `spec` and will be used declaratively within the `Panel` component in the dashboarding framework. The `Panel` component will handle plugin loading (eg lazy loading). While `TimeSeriesChart` will be decoupled from the spec and have the ability to also include richer props beyond the limits of a `Json` type (e.g. functions and event handlers). The latter will be similar to a raw React component with hooks into the context

eunicorn commented:
> > * update migration script
> > * decouple query editor from panel plugin spec **(or put the full JSON into the JSON tab like we discussed earlier today)**
> 
> Is it safe to do these follow-ups separately or are we potentially going to leave things in a broken state if someone does a release before those two get fixed? If the latter, it might be better to work this out in a feature branch, so you can balance having several PRs without shipping perses in a potential state that breaks something.

I will have other feature branches so all changes will be merged together. Here's the first follow up PR to decouple query editor and display full JSON https://github.com/perses/perses/pull/1060 :)

saminzadeh reviewed (approved):
LGTM - thanks for tackling this big one! I think it will be a good change for better composability

eunicorn commented:
@AntoineThebaud can you help take a look and see why there are still two BE tests failing? 🙏

eunicorn commented on ui/dashboards/src/components/GridLayout/GridLayout.tsx:
```diff
@@ -75,11 +83,15 @@ export function GridLayout(props: GridLayoutProps) {
-                <GridItemContent panelGroupItemId={{ panelGroupId, panelGroupItemLayoutId: i }} />
+                <GridItemContent
+                  panelGroupItemId={{ panelGroupId, panelGroupItemLayoutId: i }}
+                  width={w * gridColWidth}
```
calculating the panel width here so we get a more appropriate estimate for `useSuggestedStepMs`

juliepagano replied:
I did a little testing and I don't think the math here is quite right. The `width` you pass is what's console logged  in the screenshot, and it doesn't match the element width. I tried this with resizing a bunch to make sure I didn't hit a weird edge case.

<img width="1290" alt="image" src="https://user-images.githubusercontent.com/396962/231909035-55e5c850-901c-4e94-b395-5f70b9ad73bc.png">

It looks like `react-grid-layout` helpfully has some utils with the right math in them. Sadly they aren't exported for direct use, but hopefully you could use them to tweak the calculations you have here and in `handleWidthChange`

https://github.com/react-grid-layout/react-grid-layout/blob/master/lib/calculateUtils.js#L14-L35

eunicorn replied:
oh thanks i forgot to include the margin for each column when calculating the item width!

eunicorn commented on ui/dashboards/src/components/PanelDrawer/PanelPreview.tsx:
```diff
@@ -11,15 +11,25 @@
+import { useSuggestedStepMs } from '../../utils';
 
 const PANEL_PREVIEW_HEIGHT = 300;
+const PANEL_PREVIEW_DEFAULT_WIDTH = 800;
```
setting the default width as 800 for now since our panel editor's width is fixed

juliepagano reviewed (commented):
I mostly looked at the new grid adjustments. I'll take another pass at the rest of the code in the morning.

AntoineThebaud commented:
> @AntoineThebaud can you help take a look and see why there are still two BE tests failing? 🙏

Ok actually this is caused by the CUE version upgrade, seems they adjusted the build of some error messages:
```
         	Error:      	Not equal: 
        	            	expected: "invalid query : spec: field not allowed: aaaaaa"
        	            	actual  : "invalid query : spec.aaaaaa: field not allowed"
```
'will open a small PR to your branch to fix this

Nexucis commented:
that's a bit weird it didn't fail previously :o. Or did you improve the quality of the error message @AntoineThebaud ?

AntoineThebaud commented:
> that's a bit weird it didn't fail previously :o. Or did you improve the quality of the error message @AntoineThebaud ?

Yes this part was affected by this datamodel change (for the better, the message is much more explicit now) https://github.com/perses/perses/pull/1032/files#diff-b43d233b95a2e0c61612199d432b63914e5f77d345afd4fb4c549cab150e6400R179

AntoineThebaud commented:
It's all good now @eunicorn :)

AntoineThebaud commented on docs/cue.md:
```diff
@@ -21,8 +21,9 @@ spec: close({
 it should contain:
-- a package name.
-- the variable's `kind`.
+
+- a package name,
+- the variable's `kind`,
```
Yes that comes from one of my PRs. It seems there are not precise rules for bulleted lists, personally (try to) I use this kind of conventions https://www.pristineword.com/comma-bulleted-list/, mainly capital letters + dot, or lowercase letter + commas. Anyway not strongly attached to it so I dont mind if we change this part

juliepagano replied:
I think it's fine! I'll resolve this one.

eunicorn commented:
@sjcobb @juliepagano @cndonovan finally all tests are passing and should be good for a final review! 

FYI some changes are from other PRs that have already been reviewed and merged into this branch
- https://github.com/perses/perses/pull/1072
- https://github.com/perses/perses/pull/1060

juliepagano commented on ui/dashboards/src/components/GridLayout/GridLayout.tsx:
```diff
@@ -49,6 +52,18 @@ export function GridLayout(props: GridLayoutProps) {
+    const marginX = margin[0];
+    const marginWidth = marginX * (cols - 1); // exclude margin from container width
+    const containerPaddingWidth = containerPadding[0] * 2;
+    setGridColWidth(Math.round((containerWidth - marginWidth - containerPaddingWidth) / cols));
```
```suggestion
    setGridColWidth((containerWidth - marginWidth - containerPaddingWidth) / cols);
```

Looks like the math is still not quite right. If you remove the rounding here and put it in the other function, then the numbers match up properly.

eunicorn replied:
Interesting, when i tested, it was more accurate when I moved the rounding here instead of the other function. But i can move it back. I don't think the suggestedStepMs needs to be exact

juliepagano replied:
Ah weird. It was several pixels off with the other way around, but perfectly matched when I did it this way.

juliepagano replied:
I wish we could write an automated test for this, but I think that's kind a pain with how this code is written right now, and I don't want to add more overhead to this already sufficiently complex PR.

juliepagano commented on ui/dashboards/src/components/GridLayout/GridLayout.tsx:
```diff
@@ -88,3 +108,15 @@ export function GridLayout(props: GridLayoutProps) {
+const calculateGridItemWidth = (w: number, colWidth: number) => {
+  // 0 * Infinity === NaN, which causes problems with resize contraints
+  if (!Number.isFinite(w)) return w;
+  return colWidth * w + Math.max(0, w - 1) * DEFAULT_MARGIN;
```
```suggestion
  return Math.round(colWidth * w + Math.max(0, w - 1) * DEFAULT_MARGIN);
```
See comment above. I tested these changes locally and validated that the dom size and the value on the prop match when you swap where the rounding happens.

juliepagano commented on ui/dashboards/src/components/GridLayout/GridLayout.tsx:
```diff
@@ -49,6 +52,18 @@ export function GridLayout(props: GridLayoutProps) {
     }
   };
 
+  const handleWidthChange = (
```
Could you include a comment with a link to the relevant react-grid-layout code either here or on `calculateGridItemWidth` for reference for future us. In theory, if they change underlying calculations, something here could get out-of-sync, and it'd be nice to have a quick reference to the source-of-truth. 

https://github.com/react-grid-layout/react-grid-layout/blob/master/lib/calculateUtils.js#L14-L35

juliepagano reviewed (commented):
This is super close. Ping me when those rounding changes land, and I'll give it one more quick pass and then 👍.

juliepagano reviewed (approved):
Looks good to me! Thanks for putting so much hard work and care into moving this breaking change forward. 🏆

sjcobb commented:
Hey sorry to jump back in late! Checked out from this branch for local dev and loving having the full panel JSON in the panel edit JSON tab and how you can change panel types and not lose your query 🔥 

I did hit some weirdness when I tried duplicating a panel and then switching panel types.  I see `{"message": "bad request: code=400, message=name cannot be empty, internal=name cannot be empty"}` and `bad request: invalid panel Test 1 - 1: spec.name: field not allowed`. Although name is filled in. Also at one point I got `Error: Extension context invalidated.` error

JSON examples where I duplicated the panel:
[test_name_error_02.txt](https://github.com/perses/perses/files/11243138/test_name_error_02.txt)
[test_name_error_01.txt](https://github.com/perses/perses/files/11243139/test_name_error_01.txt)

[Recording video link here ](https://drive.google.com/file/d/1Q4HXiG3a4VLOhGjHH4WprBcpjtLp1zUA/view?usp=share_link) where I reproduced it

AntoineThebaud commented:
> Hey sorry to jump back in late! Checked out from this branch for local dev and loving having the full panel JSON in the panel edit JSON tab and how you can change panel types and not lose your query 🔥
> 
> I did hit some weirdness when I tried duplicating a panel and then switching panel types. I see `{"message": "bad request: code=400, message=name cannot be empty, internal=name cannot be empty"}` and `bad request: invalid panel Test 1 - 1: spec.name: field not allowed`. Although name is filled in. Also at one point I got `Error: Extension context invalidated.` error
> 
> JSON examples where I duplicated the panel: [test_name_error_02.txt](https://github.com/perses/perses/files/11243138/test_name_error_02.txt) [test_name_error_01.txt](https://github.com/perses/perses/files/11243139/test_name_error_01.txt)
> 
> [Recording video link here ](https://drive.google.com/file/d/1Q4HXiG3a4VLOhGjHH4WprBcpjtLp1zUA/view?usp=share_link) where I reproduced it

Indeed we can see that an unexpected `name` attribute was appended within the panel plugin spec in the examples you provided
```
"plugin": {
  "kind": "StatChart",
  "spec": {
    "name": "",
    "calculation": "LastNumber",
    ...
  }
},
```

This gets rejected by the backend validation as intended, because not expected in the plugin spec (here [StatChart](https://github.com/perses/perses/blob/main/schemas/panels/stat/stat.cue#L21))

eunicorn commented:
> Indeed we can see that an unexpected `name` attribute was appended within the panel plugin spec in the examples you provided

Fixed. For some reason, `name` was in the default json for stat chart so i removed it

Nexucis reviewed (approved):
awesome jobs @eunicorn !! Well done !

sjcobb reviewed (approved):
🚀
